### ISSUE TITLE
Improvement: Characters should pathfind through transport rides.

### ISSUE THREAD

**Issue by jummy0** [OpenRCT2/OpenRCT2#7758]
**Version:** <=0.2.0

- [x] Reproducible in RCT2 (vanilla)?
- [ ] Specific to multiplayer?

**Steps to reproduce:**
1. Create a section of the park that is only accessible by transport ride.
2. Create rides in this section of the park.
Guests will be unable to find the distant section of the park, and complain that they can't find the rides there. If a ride is broken down in the distant section, mechanics from the main section will be unable to reach the ride, but will still make an attempt. Without manual intervention, this prevents the ride from being fixed, as the single mechanic attempting to reach it locks out other mechanics that are able to do so (this is especially troublesome in multiplayer, where manually-set staff paths don't always sync between clients).

**Screenshots / Video:**
![isolated-ride](https://user-images.githubusercontent.com/4022196/42266474-1f22008c-7f3c-11e8-89bb-e2e2bb6f0772.png)

**Comment by Patrik356b** [comment]
This may cause a catch 22 with pay-per-ride, unless the transport ride price is set to free:
Guest x33 pays to go on the transport ride
Guest x33 pays to go on the isolated ride
Guest x33 is stuck because because their cash doesn't cover the transport ride fee

**Comment by Broxzier** [comment]
That's like a no-entry banner in front of the park exit, and more an issue to solve by the player. Fortunately RCT2 features ATMs.

**Comment by jummy0** [comment]
Theoretically that can happen in the current version. Guests *can* reach the isolated ride, but only if they happen to randomly take the transport. If there's any way to spend money over there they could get stuck. 

It'd be difficult to find and track a single guest that's stuck like that, unlike with staff where you might only have a handful to keep track of.

**Comment by pizza2004** [comment]
I would just solve the issue by only having free rides count as part of pathfinding.  I mean, it would be like if Disneyland expected you to pay for a ride to a specific part of the park.  You'd treat that like a ride rather than a path to somewhere else.

**Comment by duncanspumpkin** [comment]
I was sure that this was a duplicate issue but I can't find the original one. It has always been the intention to improve the pathfinding to account for transport rides rather than the current setup where there is only an increased preference to use transport rides.

**Comment by moonheart08** [comment]
@ocalhoun6 Yup, but non transport rides should probably have a higher pathfinding cost than transport rides, making guests prefer transport over rides (I want to get to my destination WITHOUT riding the entirety of Mr. Bone's wild ride, thanks) and prefer short times over high ones.

**Comment by SijmenSchoon** [comment]
Sounds really cool to me. This always disappointed me as a child - I liked making separate islands and only having transport rides between them, especially in RCT3 with the tonnes of cool transport rides.

It does deviate from the main game a bit though, so it might be better off as a plug-in, or at least as an option.

**Comment by moonheart08** [comment]
@SijmenSchoon It's a QoL feature. I'd say it'd make perfect sense in the official, as we're not stuck in the 1990s anymore.

**Comment by shuhaowu** [comment]
It looks like someone thought about doing this a long time ago, but never followed through: 113c41b97699f40236f06dd8dc7175d799baff92

I looked through the code very briefly and I feel the scope of the work is quite extensive. Apart from the source code directly, are there more technical materials written about path planning?

**Comment by MattBSG** [comment]
This would be a huge quality of life improvement if it were to be implemented and give transportation rides a true meaning and purpose. Is this on any roadmap at this time since I know it's been a while since opened?

**Comment by ocalhoun6** [comment]
It's a pretty daunting task, requiring a lot of from-the-ground-up coding. It would require a lot of rethinking how the pathfinding algorithm works, since as I understand it, peeps/staff usually don't have the entire route planned out in advance when they're navigating somewhere -- they only look a few turns ahead. And then you have to factor in things like how fast the rides move vs. how fast they walk (for different walking speeds *and* different ride times between different stations), how long the Q lines are, how long it's likely to before the next train arrives at the station, etc.

And if you're going to have staff (mechanics) riding rides to get where they're going, you'd probably need to make a bunch of new sprites/drawing code because there currently aren't any sprites available for a mechanic riding any ride. (It would be pretty ugly if the mechanic suddenly turned into a regular guest when he sat down on your miniature railroad train.) From what I understand at the moment, the sprites for drawing peeps on a ride are actually part of the *ride vehicle's* sprites, so every type of vehicle a mechanic could possibly ride would need new sprites to show him riding it.

**Comment by gsckoco** [comment]
This could be achieved rather simply with a dijkstra algorithm, and then have the transport ride have the other exits and entrances connected.

![image](https://user-images.githubusercontent.com/46426171/94672904-d94b9700-030d-11eb-9cfa-5dd6a36ad64e.png)

**Comment by AaronVanGeffen** [comment]
Ideally, the user would be able to choose between modern and classic pathfinding. Therefore, we would really like to make the pathfinding routines more modular before we rigorously change the approach. This should happen before a new pathfinding algorithm is introduced.

As for the algorithm, while Dijkstra's is a fundamental approach, it may be a bit naive. We have had an experimental PR for A* in the past (#6440). I think such a weighted method may lead to better results.

**Comment by gsckoco** [comment]
> Ideally, the user would be able to choose between modern and classic pathfinding. Therefore, we would really like to make the pathfinding routines more modular before we rigorously change the approach. This should happen before a new pathfinding algorithm is introduced.
> 
> As for the algorithm, while Dijkstra's is a fundamental approach, it may be a bit naive. We have had an experimental PR for A* in the past (#6440). I think such a weighted method may lead to better results.

I'll start working on making the current pathfinding more modular so I can start on a new pathfinding system.

**Comment by AaronVanGeffen** [comment]
Please check with @duncanspumpkin as well; I believe he had some plans for this.

**Comment by pizza2004** [comment]
I think pathfinding being modular is a good thing, but I don't know if I agree that the user themselves should be choosing.  I imagine it should either depend on the scenario, or perhaps be set on a peep by peep basis (with cheats so that the user can change it if they want).  Otherwise the high scores generated aren't going to be strictly comparable anymore.

**Comment by gsckoco** [comment]
The main rationale behind making it modular is so that it can be switched. As for it being per peep, why? Surely that would remove the whole purpose of the improved pathfinding, hiding it in some obscure cheat menu and you having to go through and enable it for each of your few thousand peeps?

**Comment by pizza2004** [comment]
I don't actually think per peep would be a good idea, I just figured if people thought it wasn't something that should be scenario specific then that was an alternative example.  The main point I wanted to make is that giving the user direct control over something like this without making a cheat is basically giving them a difficulty setting switch without any indication of what "difficulty" they're using reflected in any end result.

**Comment by sijmn** [comment]
One advantage of having per peep would be that you could make peeps with maps smarter. It wouldn't make sense if a peep without a map could just pathfind the most optimal path to a ride from anywhere.

You could also adjust it based on their needs, e.g. a tired peep could prefer transport rides over walking, even though walking would be shorter.

**Comment by sijmn** [comment]
Also, what should be kept in mind, is that the Sawyer most likely considered this exact feature, but decided not to for some reason - very likely either performance (less relevant nowadays) or game balance (still very much relevant)

**Comment by ZehMatt** [comment]
The decision was probably a bit of both, performance and balance. It would be really odd if peeps would start using transportation rides only from when they are available, but performance was probably the biggest concern back then, having 32 MiB memory pretty good when the game was developed, so caching off all possible paths was probably a bad idea back then, recomputing them all the time even worse.

**Comment by Ryder17z** [comment]
I'd say not all guests should have the same patience too, so some would value the transport ride less than others depending on queuing times and etc

**Comment by gsckoco** [comment]
> There are a few things to consider when it comes to transportation rides:
> 
> * What happens when the queue is full.
> * Should guests consider the time between waiting for the transportation ride to arrive vs walking.
> * Should guests consider the time the ride will take vs walking.
> * The cost of the ride.
> * Guests should not have god like knowledge, they must own a map.
> * Should they prefer rides when it rains?
> * How does a guest determine when to take a ride instead of walking: random, happiness, energy, etc.
> 
> The path finding is from a technical standpoint easy, keeping the game balanced not so much.

Stuff like the cost of the ride can use pre-existing calculations, and everything else would just be messing around with numbers. All of the criteria is easy to meet, sorta of the main reason of using a weighted graph.

You would increase the weight of taking the transport ride based on all of those factors.
- If the queue is full add 1000 to the weight
- If the ride costs too much add 1000 weight 
- If they don't own the map, ignore the ride from pathfinding
- You could decrease the weight of the ride when its raining, make the guest more likely to use it.
- If the guest has lots of energy, increase the weight of the ride, if the guest is low on energy decrease the weight.

**Comment by gsckoco** [comment]
The hard part is the implementation of the pathfinding, itself as opposed to the deciding factors of taking the ride

**Comment by henke37** [comment]
One question that comes to mind here is if peeps need different ride logic.

Specifically, should peeps be allowed to choose to stay on the ride so that they can use another exit station?

**Comment by ocalhoun6** [comment]
@henke37 
Oh no -- that would make things even more complicated!

Just have the peep exit the transport ride normally at each station, and then pathfind from there. If the optimal path from there means getting on the transport ride again, they'll have to go through the queue of the next station to keep going.

@ZehMatt 
Another balance issue might be that this suddenly makes transport rides much more popular ... making the transport rides constantly overwhelmed with more peeps than they can handle, especially in big parks. Some transport rides have pretty low throughput, but there could be _lots_ of peeps trying to pathfind through them. (While a path has infinite throughput. Peeps will complain about overcrowding if the path is too full, but none of them will be blocked or delayed trying to get through.)

**Comment by pizza2004** [comment]
That's a big reason I think the pathfinding should maybe depend on scenario, because older scenarios were obviously designed without this sort of thing in mind and it could drastically alter game balance, but if you designed a scenario with, say, two big areas of the part that are disconnected and very far apart with only a small amount of rented land between them (or if possible just a single transport ride that exists on area the player can't own, because that's an interesting concept, that they can't demolish the ride), then it would make sense why peeps need to use the transport ride.

For the sake of memory and processing limits I agree that it makes more sense to just have peeps get off and on, but I think this all adds to the overall challenge of making sure your transport rides are efficient (ie, many short trains instead of fewer long ones).

**Comment by iain12321** [comment]
Rather than scenario to scenarios variance, how about all guests use the original pathfinding algorithm as a default to keep all existing parks functioning the same as they do currently.

Then you can choose whether or not add a separate kiosk with a map that enables a new algorithm to your for only those guests who buy the new map.

Think this would be the best of both worlds, if you want to keep the original pathfinding just don't place the new kiosk. If you place the new kiosk and don't like it in your park, remove the enable map kiosks and shut the park, once all guests have exited you reopen it and you are back to the original pathfinding.

**Comment by sijmn** [comment]
@iain12321
Gameplay-wise, that'd be a very nice solution. I feel it's a little bit off to deviate from the original game by so much though. While adding a setting that changes/fixes path finding feels like a minor tweak, adding whole new gameplay elements feels like a major change.

**Comment by iain12321** [comment]
Yeah i definitely accept it would be a pretty big change, but do you not think the designers would have let the guests use transport as transport if it was a possibility with the hardware that was commonly available at the time though? Main argument i have read seems to be the computational power required didn't exist at the time.  I can't see it happening but would be nice.

**Comment by pizza2004** [comment]
With any change of behavior it does come down at least partially to game design instead of just "is this technically possible" and so it's really a question of whether it enriches the gameplay without removing the challenge, or possibly if it removes some challenge but creates some other challenge in turn.  That's the main reason I was saying a scenario option makes the most sense, because it would be such a drastic change to old maps that it could skew high scores or cause unintended difficulty changes that ruin older scenarios.  A new building doesn't really fix that either, since it's basically the exact same option, but instead of being part of the scenario it's part of an object that you have to add to your scenario, which basically makes it like requiring an "upgrade" to do it.

I still think just leaving the default behavior alone and only having peeps pathfind through transport rides if they are holding a map makes the most sense, and to limit it to scenarios that are specifically marked to use the new pathfinding, once we have a new save format, for the sake of making it not break older scenarios.  That way it allows the people crafting scenarios more leeway to create interesting options, instead of forcing their hand.

**Comment by iain12321** [comment]
Personally i disagree with leaving it as transport is just seen by guests as another ride, but i accept that it would alter it in a big way. Guess i need to teach myself to code then change my own game..

**Comment by tylerleamon** [comment]
Looking to possibly pick up this discussion again if it's still a requested feature to be implemented in OpenRCT2. I have a working solution that uses the current recursive pathfinding algorithm with not too many changes and it's modular enough to address the concerns in this thread. Essentially, any ride (transport and rides such as Thunder Rocks Park and Nevermore Park) is considered for pathfinding, but that can easily be subject to change. What would be the best way to continue? I would imagine some changes/balancing may be needed before I initiate a pull request.

**Comment by IntelOrca** [comment]
> This definitely would be a huge quality of life change in the ability to have split park areas without guests getting lost and finally give a true purpose to transport rides other than just generally looking nice!

Guests will ride transport rides if they are "leaving the park", so they are already partially providing more purpose than just another ride.

**Comment by tylerleamon** [comment]
> > This definitely would be a huge quality of life change in the ability to have split park areas without guests getting lost and finally give a true purpose to transport rides other than just generally looking nice!
> 
> Guests will ride transport rides if they are "leaving the park", so they are already partially providing more purpose than just another ride.

True, but that only applies if the guest is at the entrance of the transport ride itself. Current pathfinding does not direct a peep to head towards a transport ride if they are leaving the park. With my additions, the guest still uses ShouldGoOnRide() when pathfinding to follow the same "leaving the park" logic, but now they explicitly head towards the transport ride entrance if it results in a quicker path.

**Comment by MontyThibault** [comment]
@tylerleamon Did you ever finish this, or is your code at least available? I see you have a pathfind_transport branch on your forked repo but there are no relevant commits.

**Comment by tylerleamon** [comment]
I just pushed a branch of my current [pathfind_transport implementation](https://github.com/tylerleamon/OpenRCT2/tree/pathfind_transport) I'm not sure where to go from here with regards to a pull request. Below is a small writeup of my implementation.

### Design
The goal of my transport ride pathfinding solution was to refrain from modifying the existing pathfinding algorithm as much as possible, in hopes that acceptance and integration would be easier. While I did consider a rewrite of the entire pathfinding code, it was unknown how large of an effort this would result in. Either direction will still require game balancing decisions to be made using rides as transportation during pathfinding.

### Assumptions
While searching an edge, only 1 transport ride can be used per edge to successfully reach the goal. While my code does allow for the expansion of multiple rides, this is considered future work.
Penalties for using rides as transport are determined by the rides queue time and segment length.
Guests will need to remember pathfinding across saves, so adding metadata to new .park save file is necessary.
The UI should be updated to display when a guest is heading towards a ride for transport versus the goal ride.

### Implementation
Current pathfinding detects ride queues, but does not continue pathfinding and is considered a dead end. The majority of the pathfinding algorithm relies of recursive calls to peep_pathfind_heuristic_search() and updating metadata when the pathfinding goal has been achieved.

### New Transport
Instead of treating ride queues as dead ends, the pathfinding algorithm will continue to search through the queue to find the ride entrance. At this point, the new transport pathfinding takes over. The transport pathfinding with check if the peep wants to ride the transport ride. If so, the algorithm will find the exit of the the current transport ride. If an exit is found that has a connecting path, the transport pathfinding algorithm will pass in the exit path and continue the original peep_pathfind_heuristic_search() algorithm. 

### Discussions Needed

- [ ] - Determine penalties for queue time and segment length.

### Testing Needed

- [ ] New/Existing Save Files with missing metadata for new Guest transport pathfinding variables.
- [ ] Scenarios with expanded Max Guest size, all using new pathfinding algorithm and performance impact.

**Comment by ocalhoun6** [comment]
>Determine penalties for queue time and segment length.

These penalties should maybe depend on how much energy the peep has. Since peeps with more energy walk faster (sometimes _drastically_ faster), that would greatly affect whether it's slower or faster to take a transport ride instead.

It also makes sense from a real-world perspective. If you're tired, you're probably more likely to consider an easier mode of transport.

I'm assuming you're measuring queue length and segment length by time, right? Since that's available. Then you'd just need to... 

1: Take the peep's energy level, multiply that by a conversion factor to estimate their speed.

2: Take the queue time + the ride segment time, divide that by the peep's speed, and now you have the distance-equivalent of the transport ride. (You'll know, for example, that the peep could walk 150 tiles distance in the time it would take to ride that transport ride to their destination.)

3: Then you just add this distance-equivalent to the distance calculation of that pathfinding branch.

So, for example...

1: The peep has low energy, giving them a low speed of 15 path tiles per minute.

2: The queue time for the transport ride is 4 minutes, the segment length is 6 minutes. For a total time of 10 minutes. The peep could walk 150 tiles in that time.

3: One pathfinding branch finds a possible path including the transport ride: 10 tiles to the ride + 150 tile equivalent-distance to take the ride + 20 tiles after the ride exit to the peep's destination, for a total length of 180 tiles. Another pathfinding branch finds a path to get there without the transport ride, and that path would take 200 tiles. Since the branch with the transport ride is shorter, the peep takes the transport ride.

(Or really, getting closer to how it actually works, I think: After searching 10 tiles, a pathfinding branch finds the entrance to a transport ride, and -- consulting the peep's speed -- finds that the ride segment's distance-equivalent is 150 tiles. That branch of the pathfinding then pauses while all other branches of the pathfinding progress until the other branches reach 160 tiles in length. The transport ride's path is then considered in parallel with other branches again as they expand. When all of the branches have reached 180 length units, the transport ride's branch has found the peep's destination and gets selected. If it hadn't been, then the pathfinding algorithm would have kept going until it reached 200 tiles long and found the walking-only route.)

**Comment by tylerleamon** [comment]
> > Determine penalties for queue time and segment length.
> 
> These penalties should maybe depend on how much energy the peep has. Since peeps with more energy walk faster (sometimes _drastically_ faster), that would greatly affect whether it's slower or faster to take a transport ride instead.
> 
> It also makes sense from a real-world perspective. If you're tired, you're probably more likely to consider an easier mode of transport.
> 
> I'm assuming you're measuring queue length and segment length by time, right? Since that's available. Then you'd just need to...
> 
> 1: Take the peep's energy level, multiply that by a conversion factor to estimate their speed.
> 
> 2: Take the queue time + the ride segment time, divide that by the peep's speed, and now you have the distance-equivalent of the transport ride. (You'll know, for example, that the peep could walk 150 tiles distance in the time it would take to ride that transport ride to their destination.)
> 
> 3: Then you just add this distance-equivalent to the distance calculation of that pathfinding branch.
> 
> So, for example...
> 
> 1: The peep has low energy, giving them a low speed of 15 path tiles per minute.
> 
> 2: The queue time for the transport ride is 4 minutes, the segment length is 6 minutes. For a total time of 10 minutes. The peep could walk 150 tiles in that time.
> 
> 3: One pathfinding branch finds a possible path including the transport ride: 10 tiles to the ride + 150 tile equivalent-distance to take the ride + 20 tiles after the ride exit to the peep's destination, for a total length of 180 tiles. Another pathfinding branch finds a path to get there without the transport ride, and that path would take 200 tiles. Since the branch with the transport ride is shorter, the peep takes the transport ride.
> 
> (Or really, getting closer to how it actually works, I think: After searching 10 tiles, a pathfinding branch finds the entrance to a transport ride, and -- consulting the peep's speed -- finds that the ride segment's distance-equivalent is 150 tiles. That branch of the pathfinding then pauses while all other branches of the pathfinding progress until the other branches reach 160 tiles in length. The transport ride's path is then considered in parallel with other branches again as they expand. When all of the branches have reached 180 length units, the transport ride's branch has found the peep's destination and gets selected. If it hadn't been, then the pathfinding algorithm would have kept going until it reached 200 tiles long and found the walking-only route.)

Exactly! Agreed on all fronts and shouldn’t be too hard to implement. At some point I had to stop development and submit my thoughts/implementation to see if it has a way of merging into develop in the future. I chose static penalties as a starting point to get the overall discussion started.

**Comment by ocalhoun6** [comment]
Just thought of something, though...

What happens if the queue line of the transport ride is full and the peep is unable to join the queue?

**Comment by ajayyy** [comment]
For anyone still interested in this, I updated @tylerleamon 's branch to be up to date with develop https://github.com/ajayyy/OpenRCT2/tree/pathfind_transport

**Comment by ajayyy** [comment]
One problem with this is that it seems the guests will still refuse to take the transport ride if it is not thrilling enough

**Comment by ZehMatt** [comment]
We should first strive to actually refactor the current guest/peep code, right now there is way too much behavior code mixed within the path finding code, this is overall a mess currently and I highly recommend to not try to hack anything into this area of code as it stands. There have been multiple attempts by various people and it never got really any better I would even say in its current state it has gotten worse by the last attempt. I'm not saying this to to discourage people from trying to work on this subject but the expectation should be set accordingly, this part of the code is a wild mess.

**Comment by MarioPL98** [comment]
Hi, just checking if there have been any updates on this matter?

**Comment by Nrgte** [comment]
> Hi, just checking if there have been any updates on this matter?

You can check out my fork of OpenRCT2. It has pathfinding through transport rides implented:

https://github.com/Nrgte/OpenRCT2-SIM

**Comment by EliasRipley** [comment]
> > Hi, just checking if there have been any updates on this matter?
> 
> You can check out my fork of OpenRCT2. It has pathfinding through transport rides implented:
> 
> https://github.com/Nrgte/OpenRCT2-SIM

Have you tried making a pull request of just the pathfinding to the OpenRCT2 project @Nrgte ?
I've stumbled onto this thread as I was thinking of just doing the pathfinding via transport rides.

**Comment by Nrgte** [comment]
> 
> Have you tried making a pull request of just the pathfinding to the OpenRCT2 project [@Nrgte](https://github.com/Nrgte) ? I've stumbled onto this thread as I was thinking of just doing the pathfinding via transport rides.

No it deviates too much from vanilla RCT2 + the code is engangled and dependant on other features.

**Comment by EliasRipley** [comment]
> > Have you tried making a pull request of just the pathfinding to the OpenRCT2 project [@Nrgte](https://github.com/Nrgte) ? I've stumbled onto this thread as I was thinking of just doing the pathfinding via transport rides.
> 
> No it deviates too much from vanilla RCT2 + the code is engangled and dependant on other features.

Cheers for the reply.
I did actually start on this today.
I decided to do it in a way where it is a park setting 'Allow free transport rides to be used to navigate to other destinations.' That way we should be able to keep vanilla in a safe state.
